patchwork-registries/src/main/java/net/minecraftforge/registries/RegistryManager.java

```diff
@@ -43,28 +53,121 @@ public String getName() {
+
+		if (existingKey == null) {
+			return null;
+		}
+
+		return getRegistry(existingKey);
 	}
 
 	public <V extends IForgeRegistryEntry<V>> Identifier getName(IForgeRegistry<V> reg) {
 		return reg.getRegistryName();
 	}
 
-	/**
-	 * Used by {@link net.minecraftforge.registries.ForgeRegistries}.
-	 *
-	 * @param registry the registry to add to the mapping
-	 */
-	void addRegistry(Identifier key, ForgeRegistry registry) {
-		registries.put(key, registry);
+	public Identifier getName(Class<?> clazz) {
+		if (clazz == null) {
+			return null;
+		}
+
+		do {
+			Identifier existingKey = RegistryManager.ACTIVE.superTypes.get(clazz);
+
+			if (existingKey != null) {
+				return existingKey;
+			}
+
+			clazz = clazz.getSuperclass();
+		} while (clazz != null && clazz != Object.class);
+
+		return null;
+	}
+
+	public Set<Identifier> getRegistryNames() {
+		return Registry.REGISTRIES.getIds();
+	}
+
+	<V extends IForgeRegistryEntry<V>> ForgeRegistry<V> createRegistry(Identifier name, RegistryBuilder<V> builder) {
+		Set<Class<?>> parents = Sets.newHashSet();
+		findSuperTypes(builder.getType(), parents);
+		SetView<Class<?>> overlappedTypes = Sets.intersection(parents, superTypes.keySet());
+
+		if (!overlappedTypes.isEmpty()) {
+			Class<?> foundType = overlappedTypes.iterator().next();
+			throwSuperClassOverlapException(foundType, builder.getType());
+		}
+
+		ForgeRegistry<V> reg = new ForgeRegistry<V>(this, name, builder);
+		MutableRegistry vanillaReg = (MutableRegistry) reg.getVanilla();
+		Registry.REGISTRIES.add(name, vanillaReg);
+		superTypes.put(builder.getType(), name);
+
+		return getRegistry(name);
+	}
+
+	private void findSuperTypes(Class<?> type, Set<Class<?>> types) {
+		if (type == null || type == Object.class) {
+			return;
+		}
+
+		types.add(type);
+
+		for (Class<?> interfac : type.getInterfaces()) {
+			findSuperTypes(interfac, types);
+		}
+
+		findSuperTypes(type.getSuperclass(), types);
+	}
+
+	private void throwSuperClassOverlapException(Class<?> foundType, Class<?> superClazz) {
+		LOGGER.error(
+				"Found existing registry of type {} named {}, you cannot create a new registry ({}) with type {}, as {} has a parent of that type",
+				foundType, superTypes.get(foundType), name, superClazz, superClazz);
+		throw new IllegalArgumentException(
+				"Duplicate registry parent type found - you can only have one registry for a particular super type");
+	}
+
+	@SuppressWarnings("unchecked")
+	public <V extends IForgeRegistryEntry<V>> IForgeRegistry<V> wrapVanilla(Identifier identifier, Class<V> superClazz) {
+		Registry<V> registry = (Registry<V>) Registry.REGISTRIES.get(identifier);
+
+		Set<Class<?>> parents = Sets.newHashSet();
+		findSuperTypes(superClazz, parents);
+		SetView<Class<?>> overlappedTypes = Sets.intersection(parents, superTypes.keySet());
+
+		if (!overlappedTypes.isEmpty()) {
+			Class<?> foundType = overlappedTypes.iterator().next();
+
+			if (overlappedTypes.size() == 1) {
```

Comment:
Should explain the significance of this check. What's going on here? Also, how does this handle stuff like Feature/StructureFeature?